plugin/LaserGeometryRenderer.h

```diff
@@ -59,12 +55,48 @@ class LaserGeometryRenderer : public GeometryLayerUser
   int64_t computeVersion() const override
     {return (int64_t)getCurrentFrameNumber();}
 
+  void deactivate() override
+  {
+    BaseClass::deactivate();
+    LaserDevice* laserDevice = device.getDevice();
+    if (laserDevice)
+      laserDevice->updateFrame(std::vector<Point>()); // FIXME: I'm not sure to understand why we need this. Side question: does it work to have multiple renderers?
```

Comment:
> I'm not sure to understand why we need this. 

The LASER's DAC callback is called from a different thread and may be called at a different rate than that of the graphics renderering thread. In order to send frames from the graphics rendering thread to the DAC callback we use a bounded, thread-safe, non-blocking queue. The DAC callback always writes the frame that was received most recently to the output. This means that when we deactivate the renderer or mute the `LaserDevice`, we need to ensure that a blank frame was the last frame sent to the DAC callback, otherwise the DAC callback will keep writing the last frame that was received. I hope this clarifies things!

> Side question: does it work to have multiple renderers?

We have successfully tested multiple renderers targeting different `LaserDevice`s, as well as one renderer targeting multiple devices. However, currently a single `LaserDevice` is not responsible for concatenating the output of multiple different renderers. In order to support multiple renderers targeting a single `LaserDevice`, the following is required:

1. Some new method should be added to the `LaserDevice` to allow a renderer to concatenate some points onto the laser device's current "frame".
2. The renderers should call this new method instead of `updateFrame`.
3. The `LaserDevice` needs to override some method or callback that gets called once per frame *after* all of the renderers' `renderGeometry` methods have been called. At this moment, the `LaserDevice` sends the fully concatenated frame to the DAC callback via the `updateFrame` method. The `LaserDevice` then clears its local copy of the frame, ready for the next frame to be built by the renderers.

Reply:
I think you exactly summarized what needs to be done: LaserDevice should be responsible of concatenating (at every frame).
In order to do this, you can override the update function in your Device:
`  virtual void update(const FrameInformation& frame);`
I would in every update:
 - flush current concatenation of points, i.e. send it to the DAC thread
 - clear current concatenation of points
And then, every LaserGraphicsRenderer is responsible for pushing its sequence of points during its rendering.